Change defaults for mount writable and volume copying

cpuguy83 opened the pull request:
Changes the `writable` CLI flag to `readonly` instead. This way CLI users
must opt-in to readonly as is expected from `docker run`.
The API is unchanged and defaults to to not writable.

Fixes #23915

thaJeztah commented:
LGTM

ping @stevvooe @tiborvass

vdemeester commented:
> The API is unchanged and defaults to to not writable.
Doesn't this make the API less consistent with the `/container/…` one then ? :angel:

thaJeztah commented:
I actually agree with @vdemeester, IMO we should have the same defaults on the client and API. Sending an "empty" request through the API should pick the defaults we decide on. If we think "RW" is the right default for volumes, the API should use that as a default.

stevvooe commented:
@vdemeester @thaJeztah We can change it, but it is an API change during the RC period. I've been advocating this change, but there seems to be push back.

thaJeztah commented:
Changing after GA will be even harder, I think...

> On 30 Jun 2016, at 16:04, Stephen Day <notifications@github.com> wrote:
> 
> @vdemeester @thaJeztah We can change it, but it is an API change during the RC period. I've been advocating this change, but there seems to be push back.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

stevvooe commented:
@thaJeztah @cpuguy83 @aaronlehmann I've submitted docker/engine-api#298. If we decide this is what we want, as this has come up a _few_ times, merge that and I'll take care of the deltas.

cpuguy83 commented:
This is fine, the big thing is going to be changing it in swarmkit.

stevvooe commented:
@cpuguy83 I've submitted a fix in swarmkit and engine-api. Let's update the vendoring here to use this PR as a gate for the feature.

stevvooe commented:
LGTM

nishanttotla commented:
@cpuguy83 if you update `engine-api` to the latest commit, you'll also need these changes: https://github.com/nishanttotla/docker/commit/9d62e6c1f8a1f9c8aba2f0483112ffb1be079e11

In case you don't wish to mix them in this PR, don't vendor in changes introduced by https://github.com/docker/engine-api/pull/297, and I'll open a PR with my branch after this one is merged.

dnephin commented on pkg/testutil/assert/assert.go:
```diff
@@ -15,33 +18,38 @@ type TestingT interface {
+func fatal(t TestingT, msg string) {
+	_, file, line, _ := runtime.Caller(2)
+	t.Fatalf("%s:%d: %s", filepath.Base(file), line, msg)
+}
```
What's up with this `fatal()`? isn't this already what `t.Fatalf()` prints by default?

dnephin replied:
Oh nevermind , I missed this was in the `assert` package. I have a PR somewhere that adds this as well.  I'll resolve the conflict if this merges first/

cpuguy83 commented:
Thanks @nishanttotla! Extremely helpful.

Commit:
re-vendor engine-api

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

Commit:
re-vendor swarmkit

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

Commit:
Use newer default values for mounts CLI

In the API:
`Writable` changed to `ReadOnly`
`Populate` changed to `NoCopy`

Corresponding CLI options updated to:
`volume-writable` changed to `volume-readonly`
`volume-populate` changed to `volume-nocopy`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Ok, this is updated.
Also changed the `Populate` field to `NoCopy` (which matches the `-v` syntax) here as well.

engine-api and swarmkit changes are vendored in.

stevvooe commented:
LGTM

icecrime commented on api/client/service/opts.go:
```diff
@@ -176,10 +176,16 @@ func (m *MountOpt) Set(value string) error {
+			continue
+		}
+
+		if len(parts) == 1 && strings.ToLower(parts[0]) == "volume-nocopy" {
```
Sorry if I missed the debates, but what's the reasoning for renaming the flag rather than simply changing their default value?

stevvooe replied:
The volume code is already confusing. We opted to fix this now, rather than deal with intertwined logic.

tiborvass commented:
LGTM

thaJeztah commented:
There's not much documentation yet on the `--mount` behavior, which can be handled as part of https://github.com/docker/docker/issues/24277, but we probably need updates to the API (v1.24 and v1.25) docs?

vdemeester commented:
LGTM 🐸

thaJeztah commented:
moved to docs review; just the API docs, then we should be able to merge

cpuguy83 commented:
Docs updated.

vdemeester commented:
docs LGTM 🐯

thaJeztah commented:
hmf, is this a known issue in CI? https://jenkins.dockerproject.org/job/Docker-PRs/29631/console

```
21:45:12 ----------------------------------------------------------------------
21:45:12 FAIL: docker_cli_service_create_hack_test.go:15: DockerSwarmSuite.TestServiceCreateMountVolume
21:45:12 
21:45:12 [d73632282] waiting for daemon to start
21:45:12 [d73632282] daemon started
21:45:12 docker_cli_service_create_hack_test.go:33:
21:45:12     waitAndAssert(c, defaultReconciliationTimeout, func(c *check.C) (interface{}, check.CommentInterface) {
21:45:12         if task.NodeID == "" || task.Status.ContainerStatus.ContainerID == "" {
21:45:12             task = d.getTask(c, task.ID)
21:45:12         }
21:45:12         return task.NodeID != "" && task.Status.ContainerStatus.ContainerID != "", nil
21:45:12     }, checker.Equals, true)
21:45:12 github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:2011:
21:45:12     ...open github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go: no such file or directory
21:45:12 ... obtained bool = false
21:45:12 ... expected bool = true
21:45:12 
21:45:12 [d73632282] exiting daemon
21:45:29 
```

thaJeztah commented:
could be related?

cpuguy83 commented:
It's a new test, I will have to update it before we merge.

cpuguy83 commented:
Whew, that's a good test. Issue with rebase had some lines mixed up.
Pushed up fixes.

Seems CI took a vacation.

thaJeztah commented:
retest this please
